### PR TITLE
Got working the storage in the Escape Pod.

### PR DESCRIPTION
**Pull request by CtrlAltDavid01**
I could only test it by relogging and restarting the server, I could not try it with 2 players.

Edit: The storage box is now starting empty and I didn't managed to figure out how to put the starting items into it on world gen.

PS: I'm new to this code and I don't fully understand it, sorry if I broke something!

### CODE REVIEW

File: NitroxClient/Communication/Packets/Processors/BroadcastEscapePodsProcessor.cs [hunk 1]
```diff
@@ -68,10 +68,8 @@ private void AssignPlayerToEscapePod(BroadcastEscapePods packet)
 
         // Done in a coroutine because we want to wait until the escape pod has fully loaded
         // at the beginning of the game.
-        private IEnumerator SyncEscapePodGuids(BroadcastEscapePods packet)
+        private void SyncEscapePodGuids(BroadcastEscapePods packet)
         {
-            yield return new WaitUntil(() => EscapePod.main.gameObject);
```
Comment (Sunrunner37) [inline NitroxClient/Communication/Packets/Processors/BroadcastEscapePodsProcessor.cs]: Why was this removed?  Do we not want to wait until the gameobject is loaded?

File: NitroxModel/DataStructures/GameLogic/EscapePodModel.cs [hunk 2]
```diff
@@ -1,21 +1,45 @@
...
+                }
+            }
+            set
+            {
+                AssignedPlayers = value;
+            }
+        }
+        [ProtoMember(2)]
+        public string Guid { get; set; }
+        [ProtoMember(3)]
```
Comment (Sunrunner37) [inline NitroxModel/DataStructures/GameLogic/EscapePodModel.cs]: Let's space these out with returns so that they don't all run together.

File: NitroxModel/DataStructures/GameLogic/EscapePodModel.cs [hunk 3]
```diff
@@ -1,21 +1,45 @@
...
+        public Vector3 Location { get; set; }
+        [ProtoMember(4)]
+        public string FabricatorGuid { get; set; }
+        [ProtoMember(5)]
+        public string MedicalFabricatorGuid { get; set; }
+        [ProtoMember(6)]
+        public string StorageContainerGuid { get; set; }
+        [ProtoMember(7)]
+        public string RadioGuid { get; set; }
+        [ProtoIgnore]
```
Comment (Sunrunner37) [inline NitroxModel/DataStructures/GameLogic/EscapePodModel.cs]: We don't want to persist the assigned players? We have the ability to have multiple life pods so may need this.

File: NitroxServer/GameLogic/EscapePodManager.cs [hunk 4]
```diff
@@ -1,22 +1,72 @@
...
+using System.Collections.Generic;
+using ProtoBufNet;
 using NitroxModel.DataStructures.GameLogic;
+using NitroxServer.GameLogic.Items;
 using UnityEngine;
+using System.Linq;
 
 namespace NitroxServer.GameLogic
 {
+    [ProtoContract]
```
Comment (Sunrunner37) [inline NitroxServer/GameLogic/EscapePodManager.cs]: I think we'll want to put the serialized data into a new model.  This will allow us to decouple the persistence layer and not litter the Manager with serialization attributes.

File: NitroxServer/GameLogic/EscapePodManager.cs [hunk 5]
```diff
@@ -47,18 +97,33 @@ public EscapePodModel[] GetEscapePods()
...
+            const string FILTERED_WATER_DATA = "AggBhgEIABAAGAAiCFVudGFnZ2VkMiRkNDdmM2FiZi1kNzMxLTQ4MjktYWY0Yy1kNmNhMzhhODFiZjU6JDIyYjBjZTA4LTYxYzktNDQ0Mi1hODNkLWJhN2ZiOTlmMjZiMEIkM2Q0NjdiMDEtODExNC00OTExLWJmODgtYWVlMmI3M2JlMGE4SAFQAAIIBBkKFVVuaXR5RW5naW5lLlRyYW5zZm9ybRABOAoPDQAAAAAVAAAAAB0AAAAAEhQNAAAAABUAAAAAHQAAAAAlAACAPxoPDQAAgD8VAACAPx0AAIA/DgoKUGlja3VwYWJsZRABE8I+EAgAEAAYACAAKAEwATgAQAMUChBMYXJnZVdvcmxkRW50aXR5EAECEAALCgdFYXRhYmxlEAEFDQAAAAA=";
+            const string FLARE_DATA = "AggBhgEIABAAGAgiCFVudGFnZ2VkMiQ1YzE4ODExNS03NWZhLTQ2NTMtYWYxMi02ZGYzMjFhM2E0NDk6JGY1YjZlYmVkLTlhZDktNGU0Ni1iMGI0LTBmOGQ2NDY3ZTA5MUIkM2Q0NjdiMDEtODExNC00OTExLWJmODgtYWVlMmI3M2JlMGE4SAFQAAIIBBkKFVVuaXR5RW5naW5lLlRyYW5zZm9ybRABOAoPDQAAAAAVAAAAAB0AAAAAEhQNAAAAABUAAAAAHQAAAAAlAACAPxoPDQAAgD8VAACAPx0AAIA/DgoKUGlja3VwYWJsZRABE8I+EAgAEAAYACAAKAEwATgAQAMJCgVGbGFyZRABDg0AAOFEEAAYACUAAAAAFAoQTGFyZ2VXb3JsZEVudGl0eRABAhAA";
```
Comment (Sunrunner37) [inline NitroxServer/GameLogic/EscapePodManager.cs]: Instead of putting items in the escape pod, why don't we put items into the players inventory when they first join the server.  This would allow everyone to get the starting items.  If you do it that way, I think you might be able to run the routine in client side code (so you don't have crazy stuff like this here)

File: NitroxServer/Serialization/World/PersistedWorldData.cs [hunk 6]
```diff
@@ -43,6 +43,9 @@ public class PersistedWorldData
         [ProtoMember(9)]
         public GameData GameData { get; set; }
 
+        [ProtoMember(10)]
+        public EscapePodManager EscapePodManager { get; set; }
```
Comment (Sunrunner37) [inline NitroxServer/Serialization/World/PersistedWorldData.cs]: As stated above, let's create a new model.  EscapePodData.

File: NitroxModel/DataStructures/GameLogic/EscapePodModel.cs (hunk 2 above)
Comment (CtrlAltDavid01) [reply NitroxModel/DataStructures/GameLogic/EscapePodModel.cs]: Sorry, never have worked with this.

File: NitroxServer/GameLogic/EscapePodManager.cs (hunk 5 above)
Comment (CtrlAltDavid01) [reply NitroxServer/GameLogic/EscapePodManager.cs]: Great idea, didn't think of that :D

File: NitroxClient/Communication/Packets/Processors/BroadcastEscapePodsProcessor.cs (hunk 1 above)
Comment (CtrlAltDavid01) [reply NitroxClient/Communication/Packets/Processors/BroadcastEscapePodsProcessor.cs]: It didn't seem to be needed and I removed it because it only let the code continue after the InitialPlayerSync so the items could not be placed in it. I'll try to restore it and see if it works that way.

**Comment by CtrlAltDavid01** [comment]
I made the changes, but "yield return new WaitUntil(() => EscapePod.main.gameObject);" can not be put back, because it delays the time too much and the items can't be placed inside the container and I have never noticed anything it would break.

**Comment by Sunrunner37** [comment]
For the future, we may want to put some of the escape pod stuff inside the initial sync.  (Though we still need to tell other players if another escape pod spawns because the code does support multiples).

**Comment by CtrlAltDavid01** [comment]
I'll work on it tomarrow, I'll put some of the escape pod stuff inside the initial sync as you mentioned and make it if a new pod is created every player get a packet about it.

**Comment by Sunrunner37** [comment]
Btw, you've done a great job.  Looks like you had a smooth ramp up!

**Comment by bannmann** [comment]
@PurpleTentacle81 and I tested commit edc0ed02e84ecbf16f176cc86f9256e1665ced25. For each test run, we started the server from scratch (no `save.nitrox`).

We can confirm that putting an item into escape pod storage is synchronized between players. It also works when the other player is offline and joins after the item was stored. The player that stored the item can even go offline before the other joins.

However, we tested four kinds of items, and they behave differently: fishes (peeper, boomerang, airsack, garryfish), metal salvage and copper work fine, but acid mushrooms never appear for the other player. Maybe there are more items that don't work.

Another problem is that removing an item from storage is not synchronized at all. For the other player, the item is still in storage and can be removed again. It seems like the item was duplicated.

**Comment by CtrlAltDavid01** [comment]
Acid mushrooms are a different bug, if you pick up one of the and relog it disappears from your inventory, I think it's not synced at all, so the problem is not in this code.

**Comment by quimnuss** [comment]
#545 to be precise

**Comment by CtrlAltDavid01** [comment]
I just fixed the item removal so anyone can test it if they want to (It was a dumb mistake), still working on the other stuff.

**Comment by bannmann** [comment]
We tested 9a9c0afb4b2bcede3d50539ab143d5c6197d5e64 and can confirm that item removal is now working fine.

Thanks, @CtrlAltDavid01! Looking forward to see this on `master` 😃
